04-onion-routing.md

```diff
@@ -1052,25 +1054,62 @@ The erring node then generates a new key, using the key type `ammag`.
-`hmac` field with the computed HMAC.
+1. data:
+    * [`20*u32`:`htlc_hold_times`]
+    * [`210*sha256[..4]`:`truncated_hmacs`]
 
-The association between the forward and return packets is handled outside of
-this onion routing protocol, e.g. via association with an HTLC in a payment
-channel.
+The field `htlc_hold_times` contains the htlc hold time in milliseconds for each hop. The sender can use this information to score nodes on latency.
+
+The erring node puts its hold time at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum supported number of hops in a route (20).
+
+The field `truncated_hmacs` contains truncated authentication codes series for each hop, with the
+same `um` key that is used for `hmac` in the `reason` field. Regular 32 byte HMACs are
+truncated to the first 4 bytes to save space.
+
+In theory this truncation makes it possible for malicious nodes to guess the
+HMAC. However, game theory is against them because a wrong guess will get
+them penalized.
+
+The size of the field is based on the maximum number of hops in a route (20) and
+the truncated HMAC size (4 bytes). Each hop adds 20 HMACs, one for each possible
+position that the hop could be at in the path. This is necessary because only
+the sender knows the position of each hop in the path.
+
+It is not required to store all 20 * 20 = 400 HMACs. The node in position 0 needs to store an HMAC for every one of the 20 positions. For the node in position 1, the maximum remaining route length is only 19. For that position just 19 HMACs need to be kept, and so on. This makes for a total number of HMACs of 20+19+18+...+1 = 210 HMACs.
+
+The layout of the `hmacs` field is shown below. The actual format is much longer,
+but for readability the format is described as if the maximum route length would
+be just three hops.
+
+`hmac_0_2` | `hmac_0_1`| `hmac_0_0`| `hmac_1_1`| `hmac_1_0`| `hmac_2_0`
+
+`hmac_x_y` is the hmac added by node `x` (counted from the node that is
+currently handling the failure message) assuming that this node is `y` hops
+away from the erring node.
+
+Each HMAC covers the following data:
+
+* The `reason` field.
+
+* The first `y+1` hold times in `htlc_hold_times`. For example, `hmac_0_2` would cover
+  all three hold times.
+
+* `y` downstream hmacs that correspond to downstream node positions relative to
+  `x`. For example, `hmac_0_2` would cover `hmac_1_1` and `hmac_2_0`.
+
+The erring node stores its 20 HMACs at the start of the array and zeroes
+out the rest. Strictly speaking the erring node would only need to add the
+single `hmac_0_0` here, because there is no downstream data to cover. However,
+for verification efficiency at the origin node, we still require all HMACs to be
+calculated. The redundant HMACs will cover portions of the zero-initialized
+data.
+
+Finally a new key is generated, using the key type `ammagext`. This key is then
```

Comment:
Decided to derive a new key rather than continue pulling bits for chacha from the `ammag` key, to minimize the chance of somehow reusing the stream. Alternatively could use a different nonce, but I believe that would be new in lightning?

Reply:
> Decided to derive a new key rather than continue pulling bits for chacha from the ammag key, to minimize the chance of somehow reusing the stream.

How does deriving a new key do a better job at avoiding key reuse than pull from the stream? The stream is generated via ECDH between the sender and a given hop, you mix in both sources of randomness (sender + node), thereby strengthening the entropy of the stream. 

On the other-hand, the strength of this new key would depend only on the entropy sources of that intermediate node (may be missing some context, drive by review here from the GH UI diff).

Reply:
No, new key doesn't do a better job than pulling from the stream. It's just that we now have two fields to encrypt and I was thinking of the case of a programming error being made where for each of these fields, the stream is initialized from the same key producing the same bits.

I think it just comes down to personal preference, no strong argument against or in favor of a new key.